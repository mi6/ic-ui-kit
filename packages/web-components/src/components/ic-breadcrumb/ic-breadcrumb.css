:host {
  display: block;
  margin: var(--ic-space-xs) var(--ic-space-xs) var(--ic-space-xs) 0;
}

:host .breadcrumb {
  display: flex;
  align-items: center;
  white-space: nowrap;
  gap: var(--ic-space-xs);
  min-height: var(--ic-space-lg);
}

:host .chevron {
  width: var(--ic-space-lg);
  height: var(--ic-space-lg);
}

:host .chevron svg {
  color: var(--ic-architectural-400);
}

:host(:first-child) .chevron {
  display: none;
}

.back-icon svg {
  color: var(--ic-hyperlink);
}

:host(.back) {
  display: flex;
  align-items: center;
}

:host(.back) .chevron {
  display: none;
}

ic-link {
  --breadcrumb-link-display: inline-flex;
  --breadcrumb-link-align-items: center;
  --breadcrumb-link-gap: var(--ic-space-xs);
}

:host([aria-current="page"]) .current-page-container {
  display: flex;
  align-items: center;
  gap: var(--ic-space-xs);
}

:host(.collapsed-breadcrumb-wrapper) {
  margin-right: var(--ic-space-xs) !important;
  display: flex;
  align-items: center;
  gap: var(--ic-space-xs);
}

:host(.collapsed-breadcrumb-wrapper) ::slotted(.collapsed-breadcrumb) {
  border: none;
  background-color: transparent;
  color: var(--ic-hyperlink);
  text-decoration: underline;
  font-weight: var(--ic-font-weight-bold);
  cursor: pointer;
}

:host(.collapsed-breadcrumb-wrapper) ::slotted(.collapsed-breadcrumb:hover),
:host(.collapsed-breadcrumb-wrapper) ::slotted(.collapsed-breadcrumb:focus) {
  outline: var(--ic-hc-focus-outline);
  border-bottom: 0.25rem solid !important;
  margin-bottom: -0.25rem !important;
  text-decoration: none;
}

.hide,
:host(.collapsed-breadcrumb-wrapper) ::slotted(.hide) {
  display: none;
}

@media (forced-colors: active) {
  .back-icon svg {
    color: currentcolor;
  }
}

@supports (text-decoration-thickness: 25%) {
  ic-link {
    --breadcrumb-link-display: flex;
  }

  :host(.collapsed-breadcrumb-wrapper) ::slotted(.collapsed-breadcrumb:hover),
  :host(.collapsed-breadcrumb-wrapper) ::slotted(.collapsed-breadcrumb:focus) {
    text-decoration-line: underline;
    text-decoration-thickness: 25%;
    text-underline-offset: 25%;
    border-bottom: 0 !important;
    margin-bottom: 0 !important;
  }
}
