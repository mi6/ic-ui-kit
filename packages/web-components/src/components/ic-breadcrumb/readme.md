# ic-breadcrumb



<!-- Auto Generated Below -->


## Properties

| Property                 | Attribute    | Description                                            | Type      | Default     |
| ------------------------ | ------------ | ------------------------------------------------------ | --------- | ----------- |
| `current`                | `current`    | If `true`, aria-current will be set on the breadcrumb. | `boolean` | `false`     |
| `href`                   | `href`       | The URL that the breadcrumb link points to.            | `string`  | `undefined` |
| `pageTitle` _(required)_ | `page-title` | The title of the breadcrumb.                           | `string`  | `undefined` |


## Methods

### `setFocus() => Promise<void>`

Sets focus on the breadcrumb.

#### Returns

Type: `Promise<void>`




## Dependencies

### Used by

 - [ic-breadcrumb-group](../ic-breadcrumb-group)

### Depends on

- [ic-link](../ic-link)

### Graph
```mermaid
graph TD;
  ic-breadcrumb --> ic-link
  ic-breadcrumb-group --> ic-breadcrumb
  style ic-breadcrumb fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


