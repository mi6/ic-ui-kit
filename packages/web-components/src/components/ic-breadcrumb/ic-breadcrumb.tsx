import { Component, Host, Prop, h, Element, Method } from "@stencil/core";
import { IcBreadcrumbDefault } from "./ic-breadcrumb.types";

import chevronIcon from "../../assets/chevron-icon.svg";
import backIcon from "../../assets/back-icon.svg";
import { isSlotUsed } from "../../utils/helpers";

/**
 * @slot icon - Content will be rendered to the left of the breadcrumb page title.
 */
@Component({
  tag: "ic-breadcrumb",
  styleUrl: "ic-breadcrumb.css",
  shadow: {
    delegatesFocus: true,
  },
})
export class Breadcrumb {
  @Element() el: HTMLIcBreadcrumbElement;

  /**
   * If `true`, aria-current will be set on the breadcrumb.
   */
  @Prop() current?: boolean = false;

  /**
   * The URL that the breadcrumb link points to.
   */
  @Prop() href?: string;

  /**
   * The title of the breadcrumb.
   */
  @Prop() pageTitle!: string;

  /**
   * @internal If `true`, back icon will be displayed.
   */
  @Prop({ reflect: true }) showBackIcon: boolean = false;

  componentWillRender(): void {
    this.setSlottedCurrentPageClass();
  }

  /**
   * Sets focus on the breadcrumb.
   */
  @Method()
  async setFocus(): Promise<void> {
    if (this.el.shadowRoot.querySelector("ic-link")) {
      this.el.shadowRoot.querySelector("ic-link").focus();
    }
  }

  private renderDefaultBreadcrumb = (
    current: boolean,
    pageTitle: string,
    describedById: string,
    href: string
  ): IcBreadcrumbDefault => {
    const hasPageTitle =
      pageTitle !== null && pageTitle !== undefined && pageTitle !== "";

    if (current && hasPageTitle) {
      return (
        <span
          class={{
            "current-page-container": current,
          }}
        >
          {isSlotUsed(this.el, "icon") && <slot name="icon"></slot>}
          {pageTitle}
        </span>
      );
    }

    return (
      <ic-link
        href={href}
        class="breadcrumb-link"
        aria-describedby={this.showBackIcon && describedById && describedById}
      >
        {this.showBackIcon && (
          <div class="back-icon" innerHTML={backIcon}></div>
        )}
        {isSlotUsed(this.el, "icon") && <slot name="icon"></slot>}
        {pageTitle}
      </ic-link>
    );
  };

  private setSlottedCurrentPageClass = () => {
    const icLink = this.el.querySelector("ic-link");
    const currentPage = "current-page";
    if (icLink) {
      icLink.classList.remove(currentPage);
      if (this.current) {
        const hasCurrentPageClass = icLink.classList.contains(currentPage);
        if (!hasCurrentPageClass) {
          icLink.classList.add("breadcrumb-link", currentPage);
        }
      }
    }
  };

  render() {
    const { current, href, pageTitle } = this;
    const describedById = `${
      pageTitle && pageTitle.toLowerCase().replace(" ", "-")
    }-describedby`;

    const hasPageTitle =
      pageTitle !== null && pageTitle !== undefined && pageTitle !== "";
    const hasHref = href !== null && href !== undefined && href !== "";

    return (
      <Host
        class={{
          back: this.showBackIcon,
        }}
        aria-current={current && "page"}
        role="listitem"
      >
        <div class="breadcrumb">
          <span innerHTML={chevronIcon} class="chevron" aria-hidden="true" />
          {this.showBackIcon && describedById && (
            <span
              id={describedById}
              class="hide"
            >{`Back to ${pageTitle}`}</span>
          )}
          {hasPageTitle && hasHref ? (
            this.renderDefaultBreadcrumb(
              current,
              pageTitle,
              describedById,
              href
            )
          ) : (
            <slot />
          )}
        </div>
      </Host>
    );
  }
}
