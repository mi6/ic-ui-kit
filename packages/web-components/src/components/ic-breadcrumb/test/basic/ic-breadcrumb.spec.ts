import { Breadcrumb } from "../../ic-breadcrumb";
import { newSpecPage } from "@stencil/core/testing";

describe("ic-breadcrumb", () => {
  it("renders", async () => {
    const page = await newSpecPage({
      components: [Breadcrumb],
      html: `<ic-breadcrumb></ic-breadcrumb>`,
    });

    expect(page.root).toEqualHtml(`
      <ic-breadcrumb role="listitem">
        <mock:shadow-root>
          <div class="breadcrumb">
            <span aria-hidden="true" class="chevron">
              svg
            </span>
            <slot></slot>
          </div>
        </mock:shadow-root>
      </ic-breadcrumb>
    `);
  });

  it("renders with page-title and href", async () => {
    const page = await newSpecPage({
      components: [Breadcrumb],
      html: `<ic-breadcrumb page-title="foo" href="/"></ic-breadcrumb>`,
    });

    expect(page.root).toEqualHtml(`
      <ic-breadcrumb href="/" page-title="foo" role="listitem">
        <mock:shadow-root>
          <div class="breadcrumb">
            <span aria-hidden="true" class="chevron">
              svg
            </span>
            <ic-link class="breadcrumb-link" href="/">
              foo
            </ic-link>
          </div>
        </mock:shadow-root>
      </ic-breadcrumb>
    `);
  });

  it("renders with page-title, href, current", async () => {
    const page = await newSpecPage({
      components: [Breadcrumb],
      html: `<ic-breadcrumb page-title="foo" href="/" current="true"></ic-breadcrumb>`,
    });

    expect(page.root).toEqualHtml(`
      <ic-breadcrumb current="true" href="/" page-title="foo" role="listitem" aria-current="page">
        <mock:shadow-root>
          <div class="breadcrumb">
            <span aria-hidden="true" class="chevron">
              svg
            </span>
            <span class="current-page-container">
              foo
            </span>
          </div>
        </mock:shadow-root>
      </ic-breadcrumb>
    `);
  });

  it("renders with icon", async () => {
    const page = await newSpecPage({
      components: [Breadcrumb],
      html: `
      <ic-breadcrumb page-title="foo" href="/" current="true">
        <svg
          slot="icon"
          width="24"
          height="24"
          viewBox="0 0 24 24"
          fill="none"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 6.19L17 10.69V18.5H15V12.5H9V18.5H7V10.69L12 6.19ZM12 3.5L2 12.5H5V20.5H11V14.5H13V20.5H19V12.5H22L12 3.5Z"
            fill="currentColor"
          ></path>
        </svg>
      </ic-breadcrumb>`,
    });

    expect(page.root).toEqualHtml(`
      <ic-breadcrumb current="true" href="/" page-title="foo" aria-current="page" role="listitem">
        <mock:shadow-root>
          <div class="breadcrumb">
            <span aria-hidden="true" class="chevron">
              svg
            </span>
            <span class="current-page-container">
              <slot name="icon"></slot>
              foo
            </span>
          </div>
        </mock:shadow-root>
        <svg fill="none" height="24" slot="icon" viewBox="0 0 24 24" width="24" xmlns="http://www.w3.org/2000/svg">
          <path d="M12 6.19L17 10.69V18.5H15V12.5H9V18.5H7V10.69L12 6.19ZM12 3.5L2 12.5H5V20.5H11V14.5H13V20.5H19V12.5H22L12 3.5Z" fill="currentColor"></path>
        </svg>
      </ic-breadcrumb>
    `);
  });

  it("should add the 'current-page' class when current prop is true", async () => {
    const page = await newSpecPage({
      components: [Breadcrumb],
      html: "<ic-breadcrumb id='ic-breadcrumb' current=true><ic-link href='/'>Link</ic-link></ic-breadcrumb>",
    });

    await page.rootInstance.setSlottedCurrentPageClass();

    expect(
      document.querySelector("ic-breadcrumb").classList.contains("current-page")
    ).toBeTruthy;
  });

  it("should call 'setFocus' when breadcrumb is focused", async () => {
    const page = await newSpecPage({
      components: [Breadcrumb],
      html: `<ic-breadcrumb></ic-breadcrumb>`,
    });

    //Can't expect anything in this test - this is to increase code coverage only
    await page.rootInstance.setFocus().toHaveBeenCalled;
  });
});
