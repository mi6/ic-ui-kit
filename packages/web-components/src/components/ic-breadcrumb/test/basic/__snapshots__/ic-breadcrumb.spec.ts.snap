// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ic-breadcrumb should render 1`] = `
<ic-breadcrumb role="listitem">
  <template shadowrootmode="open">
    <div class="breadcrumb">
      <span aria-hidden="true" class="chevron">
        svg
      </span>
      <div class="slotted-link-container">
        <span class="link-wrapper" tabindex="0">
          <slot></slot>
        </span>
      </div>
    </div>
  </template>
</ic-breadcrumb>
`;

exports[`ic-breadcrumb should render with icon 1`] = `
<ic-breadcrumb aria-current="page" class="ic-breadcrumb-aria-current-page" current="true" href="/" page-title="foo" role="listitem">
  <template shadowrootmode="open">
    <div class="breadcrumb">
      <span aria-hidden="true" class="chevron">
        svg
      </span>
      <span class="current-page-container">
        <slot name="icon"></slot>
        foo
      </span>
    </div>
  </template>
  <svg fill="none" height="24" slot="icon" viewBox="0 0 24 24" width="24" xmlns="http://www.w3.org/2000/svg">
    <path d="M12 6.19L17 10.69V18.5H15V12.5H9V18.5H7V10.69L12 6.19ZM12 3.5L2 12.5H5V20.5H11V14.5H13V20.5H19V12.5H22L12 3.5Z" fill="currentColor"></path>
  </svg>
</ic-breadcrumb>
`;

exports[`ic-breadcrumb should render with page-title and href 1`] = `
<ic-breadcrumb href="/" page-title="foo" role="listitem">
  <template shadowrootmode="open">
    <div class="breadcrumb">
      <span aria-hidden="true" class="chevron">
        svg
      </span>
      <ic-link class="breadcrumb-link" href="/" theme="inherit">
        foo
      </ic-link>
    </div>
  </template>
</ic-breadcrumb>
`;

exports[`ic-breadcrumb should render with page-title, href, current 1`] = `
<ic-breadcrumb aria-current="page" class="ic-breadcrumb-aria-current-page" current="true" href="/" page-title="foo" role="listitem">
  <template shadowrootmode="open">
    <div class="breadcrumb">
      <span aria-hidden="true" class="chevron">
        svg
      </span>
      <span class="current-page-container">
        foo
      </span>
    </div>
  </template>
</ic-breadcrumb>
`;
